AIP-86 - Add Deadline References

ferruzzi opened the pull request:
Adds the framework for DeadlineReferences, along with a few examples and unit tests.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

ferruzzi committed:
AIP-86 - Add Deadline References

vincbeck commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -123,7 +122,71 @@ class DeadlineReference(Enum):
+        """Call the method in the enum's value with the provided kwargs."""
+        if self._fixed_dt:
+            return self._fixed_dt
+        return getattr(self, self.value)(**kwargs)
```
I might miss something but where this `self.value` is initialized?

1fanwang replied:
I think that's inherited from Enum https://github.com/python/cpython/blob/main/Lib/enum.py

vincbeck replied:
Oh it is a an enum! My bad :)

1fanwang commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -123,7 +122,71 @@ class DeadlineReference(Enum):
+    def dagrun_logical_date(self, dag_id: str) -> datetime:
+        from airflow.models import DagRun
+
+        return self._fetch_from_db(DagRun, "logical_date", dag_id=dag_id)
```
nit:
consider using attribute reference from models to avoid magic/hard-coded string as col name here getting out of sync
"logical_date" -> `DagRun.logical_date`
https://github.com/apache/airflow/blob/af06798ccc5d320e6174af9d4c7b10a7333f36e8/airflow-core/src/airflow/api/common/mark_tasks.py#L151

ferruzzi replied:
Great suggestion, thanks.  Change coming.

1fanwang commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -123,7 +122,71 @@ class DeadlineReference(Enum):
+        return self.evaluate_with()
+
+    @provide_session
+    def _fetch_from_db(self, model_class: Base, column: str, session=None, **conditions) -> datetime:
```
just for my learning:
do we need this layer of abstraction? seems like we are mostly giving one condition (dag id specific)

ferruzzi replied:
I debated that a lot myself.   If anyone feels strongly, then I can drop it and the two current cases where it is used can be reverted to session.scalar calls.  Keep in mind that this is intended to be reused by any number of future References and may or may not be more useful with those.

1fanwang replied:
no strong opinion on this
makes sense if we expect to be reusing this

ferruzzi replied:
Thanks.   I'll leave this open for others to weigh in, but consider it a non-blocking comment.

1fanwang commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -123,7 +122,71 @@ class DeadlineReference(Enum):
+            query = query.where(getattr(model_class, key) == value)
+        # This should build a query similar to:
+        # session.scalar(select(DagRun.logical_date).where(DagRun.dag_id == dag_id))
+        self.log.debug("db query: session.scalar(%s)", query)
```
super nit:
fstring

ferruzzi replied:
This time I'm not lying.  This is where the Airflow CI  forces us to not use f-strings.

![image](https://github.com/user-attachments/assets/d40e9c46-7138-4f6c-b364-888d1fa9c8eb)
![image](https://github.com/user-attachments/assets/73f7cfa3-c44b-40d2-bc30-854e8cddeebe)

1fanwang replied:
TIL - turns out this is for backwards compatibility 
https://docs.python.org/3/howto/logging.html#logging-variable-data
thanks for sharing the screenshot 
(and I should get into the habit of trying things out and building in the airflow repo myself :p)

1fanwang reviewed (approved):
Thanks for the PR, LGTM, minor comments

1fanwang commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -123,7 +122,71 @@ class DeadlineReference(Enum):
+    # will execute dagrun_logical_date() to find the dagrun's logical date.
     DAGRUN_LOGICAL_DATE = "dagrun_logical_date"
+    DAGRUN_QUEUED_AT = "dagrun_queued_at"
+    _CUSTOM_REFERENCE_BASE = "_custom_reference"
```
curious what we plan to use `_CUSTOM_REFERENCE_BASE` for

ferruzzi replied:
Mostly just for reminding me to proof-read my PRs.  :sweat:     That was left over from some testing I did where I needed a Reference that wasn't related to the dagrun before I settled on the `object.__new__` version of fixed_datetime.  I'll remove it.  Thanks for catching that.

1fanwang commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -123,7 +122,71 @@ class DeadlineReference(Enum):
+        # This should build a query similar to:
+        # session.scalar(select(DagRun.logical_date).where(DagRun.dag_id == dag_id))
+        self.log.debug("db query: session.scalar(%s)", query)
+        return session.scalar(query)
```
What's the intended behavior if no matching records found
do we need to handle existence /  check if None here?

ferruzzi replied:
I'm not sure how that would come to pass, but I suppose extra checks can't hurt.   What do you think?  If you set a deadline that is one hour after the dagrun is queued, and the system can't find the queued time, what would you expect it to do?  I'm thinking a log.warn maybe?

1fanwang replied:
agreed 
this should not happen (rare), and shouldn't be fatal since technically things didn't fail.
log warn makes sense

ferruzzi replied:
Added a whole bunch of error messaging there, along with the tests to go with it.

ferruzzi replied:
Thinking on this more, I am going to roll that back.  If no match is found then that should be a failure case.  There is no way that I can think of to recover from that and an expected alarm will not be going off later.  Changing the log messages that i added into exceptions in a coming commit.

ferruzzi commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -123,7 +122,71 @@ class DeadlineReference(Enum):
+
+    def evaluate(self):
+        """Call evaluate_with() without any conditions, because it looks strange in use that way."""
+        return self.evaluate_with()
```
Any opinions on this?   I honestly only added it because `DeadlineReference.FIXED_DATETIME(tomorrow).evaluate_with()` looked awkward, but maybe it's a bad idea to add the extra method?  I feel like the `with()` implies something is required, but maybe I'm just overthinking it.

1fanwang replied:
how about just call it `evaluate` with optional args (instead of `evaluate_with`)

ferruzzi replied:
Maybe, but then I feel like just calling it evaluate() misses the implied expectation that there usually should be a condition there?

ferruzzi replied:
Alright, I have dropped evaluate().   It was overcomplicating things without adding any real value.  We can always add it back in later if people want it.  Resolving this.

uranusjr commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -97,11 +98,9 @@ def add_deadline(cls, deadline: Deadline, session: Session = NEW_SESSION):
 
 
-class DeadlineReference(Enum):
+class DeadlineReference(LoggingMixin, Enum):
```
This class is now awfully complex as an enum, and also does what normal enums don’t generally do (contain an arbitrary value via `FIXED_DATETIME`).

Maybe we should split this into two classes like this

```python
class DefinedDeadline(Enum):
    DAGRUN_LOGICAL_DATE = ...
    DAGRUN_QUEUED_AT = ...

    def evaluate(self, **kwargs):
        ...

@dataclass
class FixedDeadline:
    at: datetime

    def evaluate(self, **kwargs):
        ...

class DeadlineReference:
    DAGRUN_LOGICAL_DATE = DefinedDeadline.DAGRUN_LOGICAL_DATE
    DAGRUN_QUEUED_AT = DefinedDeadline.DAGRUN_QUEUED_AT
    FIXED_DATETIME = FixedDeadline
```

ferruzzi replied:
That's an interesting approach, thanks for the suggestion.  Let me see what I can work out along those lines.

ferruzzi replied:
Let me know what you think about the new implementation.

ferruzzi committed:
pr fixes

- removed unused _CUSTOM_REFERENCE_BASE field
- added error messaging with unit testing
- removed magic strings from the _fetch_from_db calls.

ferruzzi committed:
Refactor into multiple classes to lighten the load on the main enum interface

- Update unit tests accordingly
- Break references into a new module deadline_reference.py to avoid cluttering deadline.py
- Roll back part of the previous commit that allowed _fetch_from_db to return datetime | None.  If there is no datetime, we need to fail

uranusjr commented on airflow-core/src/airflow/models/deadline_reference.py:
```diff
@@ -0,0 +1,162 @@
+
+    ------
+    Usage:
+    -  ------
```
```suggestion
    -------
```

Acciedntal?

ferruzzi replied:
Thanks, fixed.

uranusjr commented on airflow-core/src/airflow/models/deadline_reference.py:
I wonder if this module should be put in the task sdk instead. `_fetch_from_db` can be moved to `airflow.models.deadline` and imported inside `evaluate`.

ferruzzi replied:
I honestly have no idea?   Maybe?

ferruzzi replied:
Just asked Ash and he says you are right, I'll make the change.

```
Anything used in dag code:sdk
Anything only used to evaluate/check deadlines: core
```

ferruzzi replied:
Alright, I think I've done that.  @ashb , If you happen to get time to review that I've made it work with Task SDK correctly, Id appreciate that

ferruzzi committed:
docstring typo

ferruzzi committed:
Swap to a Task SDK implementation

ferruzzi committed:
fix import error - pick a date, any date... just not that one.

uranusjr commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -183,3 +156,36 @@ def serialize_deadline_alert(self):
+
+    if result is None:
+        message = "No matching record found in the database."
+        logger.error(message)
```
We should include a bit more context here, such of what model and field we failed to fetch from.

ferruzzi replied:
I've updated all messaging and the docstring for this method.

uranusjr commented on task-sdk/src/airflow/sdk/definitions/deadline_reference.py:
```diff
@@ -0,0 +1,136 @@
+        """Validate the provided kwargs and evaluate this deadline with the given conditions."""
+        missing_kwargs = self.required_kwargs - set(kwargs.keys())
+        if missing_kwargs:
+            raise ValueError(f"Missing required parameters: {', '.join(missing_kwargs)}")
```
```suggestion
        if missing_kwargs := {k for k in self.required_kwargs if k not in kwargs}:
            raise ValueError(f"Missing required parameters: {', '.join(missing_kwargs)}")
```

ferruzzi replied:
I've added more robust kwarg verification both ways (missing and extra), thanks for the suggestion

uranusjr commented on task-sdk/src/airflow/sdk/definitions/deadline_reference.py:
```diff
@@ -0,0 +1,136 @@
+        """Evaluate this deadline with no parameters."""
+        if self.requires_conditions:
+            raise AttributeError("This deadline requires additional conditions, use evaluate_with() instead.")
+        return self.evaluate_with()
```
I feel this function is not necessary. We could just always use `evaluate_with()` directly. Probably not as pretty, but this is not supposed to be user-facing anyway.

ferruzzi replied:
Yeah, it felt really strange using `reference.evaluate_with()` without anything "with" but I'm likely just overthinking that.  I could drop it and make things more streamlined if you really don't like it.

ferruzzi replied:
Responding to your comment below about using "context" there, maybe for now I'll try just dropping `evaluate()` and leave `evaluate_with()` as it is (without "context") and see how we feel about that.

uranusjr reviewed (commented):
Another approach that may improve the evaluation interface may be to make the kwargs a required dict argument instead.

```python
# Instead of...
reference.evaluate_with()

# This might read better?
reference.evaluate_with(context={})

# Makes supplying values a bit verbose though.
reference.evaluate_with(context={"foo": 123})
```

ferruzzi committed:
Remove evaluate() in favor of only supporting evaluate_with()

- also added some parameter validation and adjusted the tests accordingly

ferruzzi committed:
improve _fetch_from_db messaging and docstring

uranusjr commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -183,3 +156,52 @@ def serialize_deadline_alert(self):
+    """
+    query = select(model_reference)
+
+    for key, value in conditions.items():
+        query = query.where(getattr(model_reference.class_, key) == value)
```
```suggestion
    query = query.where(*(getattr(model_reference.class_, k) == v for for k, v in conditions.items()))
```

ferruzzi replied:
Do you feel strongly about this one?  I find your version harder to read for some reason, but they are functionally equivalent so I can do it if you want.

ramitkataria replied:
I also think your version is easier to read @ferruzzi

uranusjr commented on task-sdk/src/airflow/sdk/definitions/deadline_reference.py:
```diff
@@ -0,0 +1,128 @@
+
+    def evaluate_with(self, **kwargs: Any) -> datetime:
+        """Validate the provided kwargs and evaluate this deadline with the given conditions."""
+        filtered_kwargs = {k: kwargs[k] for k in self.required_kwargs if k in kwargs}
```
```suggestion
        filtered_kwargs = {k: v for k, v in kwargs.items() if k in self.required_kwargs}
```

I think this is equivalent and easier to read?

ferruzzi replied:
Yeah, should be functionally identical.  I don't feel strongly either way, I'll go with yours.

ferruzzi committed:
small refactor

ferruzzi committed:
remove caplog test per email discussion

ferruzzi committed:
fix old import path

ferruzzi committed:
improve type hints

uranusjr commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -100,35 +102,6 @@ def add_deadline(cls, deadline: Deadline, session: Session = NEW_SESSION):
         session.add(deadline)
 
 
-class DeadlineReference(Enum):
```
Do we need to leave an import in this module for compatibility?

ferruzzi replied:
I don't think so.  This was added it in the previous PR purely as a placeholder so I could give an example of how something else will be used and the note on L107 made that pretty clear.  Nobody should have been using it for anything, I think it's safe to just drop it?

uranusjr commented on airflow-core/src/airflow/models/deadline.py:
```diff
@@ -186,3 +159,52 @@ def serialize_deadline_alert(self):
+    try:
+        result = session.scalar(query)
+    except SQLAlchemyError as e:
+        logger.error("Database query failed: (%s)", str(e))
```
```suggestion
        logger.exception("Database query failed")
```

Would this be more useful?

ferruzzi replied:
That's actually pretty great.   I'll be honest, I hadn't realized log.exception worked that way.  I'll get it done, thanks!

ferruzzi replied:
Is there a similar way to simplify the block below that which is currently:

```
if result is None:
    message = f"No matching record found in the database for query:\n    {pretty_query}"
    logger.error(message)
    raise ValueError(message)
```        

I always felt python needed a "log then raise" method, but maybe there has been one all along?

uranusjr commented on task-sdk/src/airflow/sdk/definitions/deadline_reference.py:
Would it be a good idea to name this just `deadline`? This would match the core module, is shorter but still clear enough (I think).

Also, it’s probably a good idea to expose DeadlineReference to `airflow.sdk` without needing to do subpackage import. It is a user-facing class, right?

ferruzzi replied:
For the first one, I think I maybe should move `models.deadline.DeadlineAlert` into the SDK as well since it is a direct user interface, which means it looks like I may have other use-cases where I want stuff in that path, so that works.  I'll rename the module.

For the second comment,  I'm sorry, I need some more context.  I don't really follow what you are suggesting.

ferruzzi committed:
re-shuffle locations
